### PR TITLE
[Storage] Fix for inconsistent casing in `x-ms-meta` prefix extraction in response handler

### PR DESCRIPTION
**Pull request by weirongw23-msft**
The gist is if Azure Python SDK receives a weirdly-cased metadata header `X-Ms-Meta-Example`, the existing code will result in a `ClientAuthenticationError`. However, any variation of `x-ms-` followed by any casing of `meta-...` will work. This PR addresses that issue.

Note that the issue lies in the request phase and not the response phase, because all headers returned by the server-side will have `x-ms-...` properly formatted to lowercase.

Test and recording is included as part of the reproduction and validation step.

### CODE REVIEW

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared/request_handlers.py [hunk 1]
```diff
@@ -4,11 +4,11 @@
...
 # --------------------------------------------------------------------------
 
-from typing import Dict, Optional
-
 import logging
-from os import fstat
+import re
 import stat
+from typing import Any, Dict, Optional
+from os import fstat
```
Comment (weirongw23-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_shared/request_handlers.py:11]: TODO: Flip the ordering here.

**Review by vincenttran-msft — COMMENTED** [review]
Great work on this! Let's wait for Jacob to take a pass before making any decisions on propagating to the rest of `_shared`, but code LGTM 😄 

Edit: Upon closer review, I am not exactly sure if your proposed solution is equivalent? We will discuss more offline.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared/request_handlers.py [hunk 2]
```diff
@@ -137,13 +137,18 @@ def validate_and_format_range_headers(
...
 
 
-def add_metadata_headers(metadata=None):
-    # type: (Optional[Dict[str, str]]) -> Dict[str, str]
-    headers = {}
-    if metadata:
-        for key, value in metadata.items():
-            headers[f'x-ms-meta-{key.strip()}'] = value.strip() if value else value
-    return headers
+def add_metadata_headers(metadata: Optional[Dict[str, str]] = None, **kwargs: Dict[str, Any]) -> Dict[str, str]:
```
Comment (weirongw23-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_shared/request_handlers.py:140]: TODO: Revert this change - `x-ms-` is required to be lowercased. This is ultimately up to the service.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared/request_handlers.py [hunk 3]
```diff
@@ -4,12 +4,12 @@
 # license information.
 # --------------------------------------------------------------------------
 
-from typing import Dict, Optional
-
 import logging
-from os import fstat
+import re
```
Comment (weirongw23-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_shared/request_handlers.py:8]: Not super important but just thought to fix the import ordering while I was here.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared/response_handlers.py [hunk 4]
```diff
@@ -60,9 +61,13 @@ def normalize_headers(headers):
 
 def deserialize_metadata(response, obj, headers):  # pylint: disable=unused-argument
     try:
-        raw_metadata = {k: v for k, v in response.http_response.headers.items() if k.startswith("x-ms-meta-")}
+        raw_metadata = {
+            k: v for k, v in response.http_response.headers.items() if re.search("x-ms-meta-", k, re.IGNORECASE)
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_shared/response_handlers.py:65]: We should use `re.compile` outside of the loops.

Comment (weirongw23-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_shared/response_handlers.py]: Yessir, and we should also make the pattern a raw string to be consistent with best Python practices.

**Review by jalauzon-msft — APPROVED** [review]
